homeassistant/components/aemet/weather_update_coordinator.py

```diff
@@ -0,0 +1,652 @@
+
+    def _convert_forecast_day(self, date, day):
+        condition = self._get_condition_day(day)
+        if not condition:
+            return None
+
+        return {
+            ATTR_FORECAST_CONDITION: condition,
+            ATTR_FORECAST_PRECIPITATION_PROBABILITY: self._get_precipitation_prob_day(
+                day
+            ),
+            ATTR_FORECAST_TEMP: self._get_temperature_day(day),
+            ATTR_FORECAST_TEMP_LOW: self._get_temperature_low_day(day),
+            ATTR_FORECAST_TIME: dt_util.as_utc(date),
+            ATTR_FORECAST_WIND_SPEED: self._get_wind_speed_day(day),
+            ATTR_FORECAST_WIND_BEARING: self._get_wind_bearing_day(day),
+        }
+
+    def _convert_forecast_hour(self, date, day, hour):
+        condition = self._get_condition(day, hour)
+        if not condition:
+            return None
+
+        forecast_dt = date.replace(hour=hour, minute=0, second=0)
+
+        return {
+            ATTR_FORECAST_CONDITION: condition,
+            ATTR_FORECAST_PRECIPITATION: self._calc_precipitation(day, hour),
+            ATTR_FORECAST_PRECIPITATION_PROBABILITY: self._calc_precipitation_prob(
+                day, hour
+            ),
+            ATTR_FORECAST_TEMP: self._get_temperature(day, hour),
+            ATTR_FORECAST_TIME: dt_util.as_utc(forecast_dt),
+            ATTR_FORECAST_WIND_SPEED: self._get_wind_speed(day, hour),
+            ATTR_FORECAST_WIND_BEARING: self._get_wind_bearing(day, hour),
+        }
+
+    def _calc_precipitation(self, day, hour):
+        """Calculate the precipitation."""
+        rain_value = self._get_rain(day, hour)
+        if not rain_value:
+            rain_value = 0
+
+        snow_value = self._get_snow(day, hour)
+        if not snow_value:
+            snow_value = 0
+
+        if round(rain_value + snow_value, 1) == 0:
+            return None
+        return round(rain_value + snow_value, 1)
+
+    def _calc_precipitation_prob(self, day, hour):
+        """Calculate the precipitation probability (hour)."""
+        rain_value = self._get_rain_prob(day, hour)
+        if not rain_value:
+            rain_value = 0
+
+        snow_value = self._get_snow_prob(day, hour)
+        if not snow_value:
+            snow_value = 0
+
+        if rain_value == 0 and snow_value == 0:
+            return None
+        return max(rain_value, snow_value)
+
+    @staticmethod
+    def _get_condition(day_data, hour):
+        """Get weather condition (hour) from weather data."""
+        val = get_forecast_hour_value(day_data[AEMET_ATTR_SKY_STATE], hour)
+        if val:
+            return format_condition(val)
+        return None
+
+    @staticmethod
+    def _get_condition_day(day_data):
+        """Get weather condition (day) from weather data."""
+        val = get_forecast_day_value(day_data[AEMET_ATTR_SKY_STATE])
+        if val:
+            return format_condition(val)
+        return None
+
+    @staticmethod
+    def _get_humidity(day_data, hour):
```

Comment:
This class is large with a lot of repetition.  Does it make sense to have a single `get_value` or maybe have a few cases like `get_numeric_value`?  Why is the upstream library not supplying numeric values for numeric data? This seems more ideal, and would eliminate the majority need to separate the numeric data from non-numeric data entirely.  We wouldn't need to check in the home assistant code in this case and only a handful of methods would be required.

Reply:
The upstream library is not supplying numeric values in numeric data because the API is not doing it, and the upstream library was implemented with the minimum effort to get this working.

Reply:
This is more of a style comment, so less critical.  The handling here makes the resulting platform code simpler too.  As you look into splitting into multiple coordinators consider ways to simplify is all.

Reply:
Ok, I'll mark this as resolved and try to simplify it when implementing multiple coordinators...
The problem is that I've got no idea on how to do it right now...

Reply:
As I said this was more style feedback, and probably differs based on personal preference.  I think it is fine as is.